### PR TITLE
WebsocketProviderV2 updates + formatting middleware bugfix

### LINKED ISSUE #3093: WebsocketProviderV2 field 'extraData'
I've encountered an issue with the web3.py. My code works fine with Ethereum contracts, but I face a ValueError about the HexBytes value being 280 bytes (expected 32) when working with contracts on the BSC network.

Name: web3
Version: 6.9.0

Error traceback:
```
Exception has occurred: ValueError
Could not format invalid value '0xd88301020a846765746888676f312e32302e34856c696e7578000000b19df4a2f8b5830cefffb860a1a315d2b3cb4db96507221c47f66e3e811db45aad4905d0950895c104cbeffa87eb4c8d2bcd1275835a08951dbeded9126c2c7985c9b5e597bde03d48c1d10325eda14b239e219baa9953384ade5f96899162cf6c52aa9d6ce2865e11617d58f84c8401e17177a03afff5ba177d1fcd859bff6a940b0e43551b38d5c85cb63e605baeeebccb0e6c8401e17178a06f53ccf029bd4c1451bbd8b956f264d13a312290307206861574b4c025a384928077cd060dd0ff0e42bb0981534e5ba1486fec61c24509d6da3e73630e2854104551e27360ac5196e18fd3665773728d98531a4cb6b9e4d628d6dcab406bceda0600' as field 'extraData'
ValueError: The value HexBytes('0xd88301020a846765746888676f312e32302e34856c696e7578000000b19df4a2f8b5830cefffb860a1a315d2b3cb4db96507221c47f66e3e811db45aad4905d0950895c104cbeffa87eb4c8d2bcd1275835a08951dbeded9126c2c7985c9b5e597bde03d48c1d10325eda14b239e219baa9953384ade5f96899162cf6c52aa9d6ce2865e11617d58f84c8401e17177a03afff5ba177d1fcd859bff6a940b0e43551b38d5c85cb63e605baeeebccb0e6c8401e17178a06f53ccf029bd4c1451bbd8b956f264d13a312290307206861574b4c025a384928077cd060dd0ff0e42bb0981534e5ba1486fec61c24509d6da3e73630e2854104551e27360ac5196e18fd3665773728d98531a4cb6b9e4d628d6dcab406bceda0600') is 280 bytes, but should be 32

The above exception was the direct cause of the following exception:

  File "test.py", line 13, in new_blocks
    async for response in w3.listen_to_websocket():
  File "test.py", line 22, in <module>
    asyncio.run(new_blocks())
ValueError: Could not format invalid value '0xd88301020a846765746888676f312e32302e34856c696e7578000000b19df4a2f8b5830cefffb860a1a315d2b3cb4db96507221c47f66e3e811db45aad4905d0950895c104cbeffa87eb4c8d2bcd1275835a08951dbeded9126c2c7985c9b5e597bde03d48c1d10325eda14b239e219baa9953384ade5f96899162cf6c52aa9d6ce2865e11617d58f84c8401e17177a03afff5ba177d1fcd859bff6a940b0e43551b38d5c85cb63e605baeeebccb0e6c8401e17178a06f53ccf029bd4c1451bbd8b956f264d13a312290307206861574b4c025a384928077cd060dd0ff0e42bb0981534e5ba1486fec61c24509d6da3e73630e2854104551e27360ac5196e18fd3665773728d98531a4cb6b9e4d628d6dcab406bceda0600' as field 'extraData'

```

my code:
```
import asyncio
from web3 import AsyncWeb3
from web3.providers import WebsocketProviderV2

async def new_blocks(): #                                           
    async with AsyncWeb3.persistent_websocket(WebsocketProviderV2("wss://bsc.getblock.io/mainnet/")) as w3:
        subscription_id = await w3.eth.subscribe("newHeads") #                  ^^^ or any BSC RPC
        try:
            while True:
                async for response in w3.listen_to_websocket():
                    print(response)
         
        except KeyboardInterrupt:
            await w3.eth.unsubscribe(subscription_id)

if __name__ == "__main__":
    asyncio.run(new_blocks())

```

### PR DESCRIPTION
**Pull request by fselmo**
### What was wrong?

- Closes #3093
- `WebsocketProviderV2` needed tests for subscriptions. While we revamp the test suite and try to get some actual integration tests for `eth_subscribe`, introduce some static tests to make sure our formatters and middleware work appropriately for subscriptions.
- Adds a timeout around trying to match the response `id` to the request `id` for a request. Also improves on the logic around matching the ids, simplifying it.
- Add unit tests for the raw response cache process. 
  - Ensure we are caching any undesired responses before returning the desired response.
  - Ensure we return the desired cached response, if cached, and that we never call recv() in that case. 
  - Test the timeout if a response with matching `id` never comes (see above bullet point for test case).
- **(breaking change for beta provider)** Refactor to a proper AsyncIterator pattern for receiving a persistent message stream from the socket.
- Update docs based on the above

### Todo:
- [x] Add entry to the [release notes](https://github.com/ethereum/web3.py/blob/main/newsfragments/README.md)

### CODE REVIEW

File: docs/providers.rst [hunk 1]
```diff
@@ -270,16 +270,15 @@ asynchronous context manager, can be found in the `websockets connection`_ docs.
...
+        ...             # handle responses here
         ...
-        ...                 if some_condition:
-        ...                     # unsubscribe from new block headers
-        ...                     unsubscribed = await w3.eth.unsubscribe(subscription_id)
-        ...                     break
+        ...             if some_condition:
+        ...                 # unsubscribe from new block headers and break out of
+        ...                 # iterator
+        ...                 unsubscribed = await w3.eth.unsubscribe(subscription_id)
```
Comment (reedsa) [inline docs/providers.rst]: Any reason to set the return value here to `unsubscribed`? Now that the wile loop is gone, perhaps dont need this variable.
```suggestion
        ...                await w3.eth.unsubscribe(subscription_id)
```

File: web3/_utils/module_testing/persistent_connection_provider.py [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+    )
+    async def test_async_eth_subscribe_mocked(
+        self,
+        async_w3: "_PersistentConnectionWeb3",
+        subscription_params: Tuple[Any, ...],
+        ws_subscription_response: Dict[str, Any],
+        expected_formatted_result: Any,
+    ) -> None:
+        sub_id = await async_w3.eth.subscribe(*subscription_params)
+        assert sub_id is not None
```
Comment (reedsa) [inline web3/_utils/module_testing/persistent_connection_provider.py]: Nit: should we keep this assertion or add it to the test below? Looks like `is_hexstr` returns False for None anyway.

**Review by reedsa — COMMENTED** [review]
Remaining nits, looking good otherwise!

File: docs/providers.rst (hunk 1 above)
Comment (fselmo) [reply docs/providers.rst]: Good eye 👌🏼

File: web3/providers/websocket/websocket_v2.py [hunk 3]
```diff
@@ -158,40 +157,52 @@ async def make_request(self, method: RPCEndpoint, params: Any) -> RPCResponse:
...
+            # "bad" cases somewhat gracefully.
+            return await asyncio.wait_for(
+                _match_response_id_to_request_id(), self.call_timeout
+            )
+        except asyncio.TimeoutError:
+            raise TimeExhausted(
+                f"Timed out waiting for response to request id `{request_id}` after "
+                f"{self.call_timeout} seconds. This is likely due to the provider not "
+                "issuing a response with the same id that was sent in the request, "
+                "which is required by the JSON-RPC spec."
```
Comment (reedsa) [inline web3/providers/websocket/websocket_v2.py]:
```suggestion
                f"Timed out waiting for response with request id `{request_id}` after "
                f"{self.call_timeout} seconds. This is likely due to the provider not "
                "returning a response with the same id that was sent in the request, "
                "which is required by the JSON-RPC spec."
```

File: web3/providers/websocket/websocket_v2.py [hunk 4]
```diff
@@ -67,7 +66,7 @@ def __init__(
         self,
         endpoint_uri: Optional[Union[URI, str]] = None,
         websocket_kwargs: Optional[Dict[str, Any]] = None,
-        call_timeout: Optional[int] = None,
+        call_timeout: Optional[float] = DEFAULT_PERSISTENT_CONNECTION_TIMEOUT,
```
Comment (reedsa) [inline web3/providers/websocket/websocket_v2.py:69]: The timeout may be `None` or could be the max float value, `wait_for` will in turn wait forever (or until the timeout of the large value provided). Since this would be intentionally set by the user, I think it is acceptable to let the connection stay open until the response id is returned, if ever. What do you think?

Comment (fselmo) [reply web3/providers/websocket/websocket_v2.py:69]: Hmm. That's a good point. We use this not only for call timeouts but for the while loop timeout. What if we set the while loop timeout to `self.call_timeout or 20` so that if it is set to `None` for call timeouts, the while loop will still max out in `20` seconds. Or whatever we decide that default is. Since it differs from the timeout for each call, I think that might be a good alternative. Thoughts?

Comment (reedsa) [reply web3/providers/websocket/websocket_v2.py:69]: Makes sense to have the timeout limit on the while loop. Paraphrasing what you said to make sure I have it right in my head, use the `call_timeout` unless it exceeds the threshold (20s or something). If `call_timeout` is larger than the threshold, the while loop will still  short circuit as desired.

Comment (fselmo) [reply web3/providers/websocket/websocket_v2.py:69]: My suggestion would've only worked for `None` but I think yours works well.

```py
timeout = self.call_timeout if self.call_timeout and self.call_timeout <= 20 else 20
```

This also prevents it from being `0` in the `if` check. Anything from `0.1` to `20` is fair game.

**Review by pacrob — APPROVED** [review]
lgtm!

File: web3/providers/websocket/websocket_v2.py [hunk 5]
```diff
@@ -158,40 +157,55 @@ async def make_request(self, method: RPCEndpoint, params: Any) -> RPCResponse:
...
+        try:
+            # Enters a while loop, looking for a response id match to the request id.
+            # If the provider does not give responses with matching ids, this will
+            # hang forever. The JSON-RPC spec requires that providers respond with
+            # the same id that was sent in the request, but we need to handle these
+            # "bad" cases somewhat gracefully.
+            timeout = (
+                self.call_timeout
+                if self.call_timeout and self.call_timeout <= 20
+                else 20
```
Comment (reedsa) [inline web3/providers/websocket/websocket_v2.py:199]: Should use the `DEFAULT_PERSISTENT_CONNECTION_TIMEOUT` constant here.
```suggestion
                if self.call_timeout and self.call_timeout <= DEFAULT_PERSISTENT_CONNECTION_TIMEOUT
                else DEFAULT_PERSISTENT_CONNECTION_TIMEOUT
```

Comment (fselmo) [reply web3/providers/websocket/websocket_v2.py:199]: Isn't this an entirely different bit of logic though? What if we agree to change the default timeout? I think that was the point of hard-coding this here.

Let's say I don't want _each_ call to the provider to take more than, say, 5 seconds. I wonder if I'm subscribed to a lot of subscriptions, if the response to my call won't be bogged down by the latency behind so many subscriptions where it may actually take longer than a call timeout to receive it.

Thoughts there?

Comment (reedsa) [reply web3/providers/websocket/websocket_v2.py:199]: That's fine then, I have no concerns with using the hardcoded value as is. Definitely would be overkill to create yet another constant just for this logic.

**Review by reedsa — APPROVED** [review]
🚀 :shipit:

File: web3/providers/websocket/websocket_v2.py (hunk 5 above)
Comment (fselmo) [reply web3/providers/websocket/websocket_v2.py:199]: Yeah in the same way I think using `self.call_timeout` feels a bit disingenuous to use here too but at least the user can tweak that to make sense if desired. Definitely open to re-thinking this a bit more before ripping off the `beta` tag.
